zerver/lib/timestamp.py

```diff
@@ -40,3 +41,97 @@ def timestamp_to_datetime(timestamp: float) -> datetime.datetime:
+
+        # time-secfrac    = "." 1*DIGIT
+        time_secfrac: str = r"\.[0-9]+"
+
+        # time-numoffset  = ("+" / "-") time-hour ":" time-minute
+        time_numoffset: str = f"[+-](?P<offset_hour>{time_hour}):(?P<offset_minute>{time_minute})"
+
+        # time-numoffset  = ("+" / "-") time-hour ":" time-minute
+        time_offset: str = f"[Zz]|({time_numoffset})"
+
+        # partial-time    = time-hour ":" time-minute ":" time-second [time-secfrac]
+        partial_time: str = f"(?P<hour>{time_hour}):(?P<minute>{time_minute}):(?P<second>{time_second})(?P<secfrac>{time_secfrac})?"
+
+        # full-date       = date-fullyear "-" date-month "-" date-mday
+        full_date: str = f"(?P<year>{date_fullyear})-(?P<month>{date_month})-(?P<day>{date_mday})"
+
+        # full-time       = partial-time time-offset
+        # full_time: str = f"(?P<time>{partial_time})(?P<offset>{time_offset})"
+        full_time: str = f"(?P<time>{partial_time})(?P<offset>{time_offset})"
+
+        # date-time       = full-date "T" full-time
+        date_time: str = f"^(?P<date>{full_date})[Tt]({full_time})$"
+
+    def __init__(self) -> None:
+        self.pattern = re.compile(self.patterns.date_time)
+
+    def extract_datetime(self, timestamp: str) -> datetime.datetime:
+        match = self.pattern.match(timestamp)
+        if match is None:  # nocoverage
+            raise ValueError("Invalid format. The string \"{timestamp}\" Does not follow RFC 3339.")
+
+        second = int(match["second"])
+        if second == 60:
+            # Python's datetime library doesn't support leap seconds so we
+            # will handle this by rounding down by one second if needed.
+            # https://docs.python.org/3.6/library/datetime.html#available-types
+            second = 59
+
+        secfrac = match["secfrac"]
+        microsecond = 0
+        if secfrac:
+            microsecond = int(float(match["secfrac"]) * 10**6)
+
+        factor = 1
+        tz_offset = match["offset"]
+        if tz_offset in ["Z", "z"]:
+            hours = 0
+            minutes = 0
+        else:
+            if tz_offset.startswith("-"):
+                factor = -1
+            hours = int(match["offset_hour"])
+            minutes = int(match["offset_minute"])
+        tz = datetime.timezone(factor * datetime.timedelta(hours=hours, minutes=minutes))
+
+        return datetime.datetime(
+            year=int(match["year"]),
+            month=int(match["month"]),
+            day=int(match["day"]),
+            hour=int(match["hour"]),
+            minute=int(match["minute"]),
+            second=second,
+            microsecond=microsecond,
+            tzinfo=tz,
+        )
+
+    def encode_datetime(self, dt: datetime.datetime) -> str:
+        # ISO 8601 strings are compatible with RFC 3339.
+        timestamp = dt.isoformat()
+        if timestamp.endswith("+00:00"):
+            timestamp = timestamp[:-6] + "Z"
+        return timestamp
```

Comment:
Is this something we can't just do via pytz?  I don't really want to be maintaining a giant block of timezone-conversion code if we don't need to.

(Also, we get to pick what date format to use in Zulip APIs, right?  Seems like it might be better to just use UNIX timestamps if so like we do in other places)

Reply:
Sure then, let's use Unix timestamps here to keep things consistent.